File: mobile/android/focus-android/app/src/androidTest/java/org/mozilla/focus/helpers/TestHelper.java
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.helpers;

import android.content.Context;
import android.content.res.Resources;
import androidx.annotation.NonNull;
import androidx.test.InstrumentationRegistry;
import androidx.test.espresso.ViewInteraction;
import androidx.test.uiautomator.UiDevice;
import androidx.test.uiautomator.UiObject;
import androidx.test.uiautomator.UiObjectNotFoundException;
import androidx.test.uiautomator.UiSelector;
import android.text.format.DateUtils;
import android.util.DisplayMetrics;

import org.mozilla.focus.R;
import org.mozilla.focus.utils.AppConstants;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;

import okhttp3.mockwebserver.MockResponse;
import okio.Buffer;
import okio.Okio;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static androidx.test.espresso.web.sugar.Web.onWebView;
import static android.view.KeyEvent.KEYCODE_ENTER;
import static junit.framework.Assert.assertTrue;
import static org.hamcrest.Matchers.allOf;

// This test visits each page and checks whether some essential elements are being displayed
public final class TestHelper {

    public static UiDevice mDevice =  UiDevice.getInstance(InstrumentationRegistry.getInstrumentation());
    public static final long waitingTime = DateUtils.SECOND_IN_MILLIS * 4;
    public static final long webPageLoadwaitingTime = DateUtils.SECOND_IN_MILLIS * 15;

    public static String getAppName() {
        return InstrumentationRegistry.getInstrumentation()
                .getTargetContext().getPackageName();
    }

    // wait for web area to be visible
    public static void waitForWebContent() {
        assertTrue(geckoView.waitForExists(waitingTime));
    }

    /********* First View Locators ***********/
    public static UiObject firstSlide = mDevice.findObject(new UiSelector()
            .text("Power up your privacy")
            .enabled(true));
    public static UiObject secondSlide = mDevice.findObject(new UiSelector()
            .text("Your search, your way")
            .enabled(true));
    public static UiObject thirdSlide = mDevice.findObject(new UiSelector()
            .text("Add shortcuts to your home screen")
            .enabled(true));
    public static UiObject lastSlide = mDevice.findObject(new UiSelector()
            .text("Make privacy a habit")
            .enabled(true));
    public static UiObject nextBtn = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/next")
            .enabled(true));
    public static UiObject finishBtn = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/finish")
            .enabled(true));
    public static UiObject initialView = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/backgroundView")
            .enabled(true));

    /********* Main View Locators ***********/
    public static ViewInteraction menuButton = onView(
            allOf(withId(R.id.menuView),
                    isDisplayed()));

    /********* Web View Locators ***********/
    public static UiObject browserURLbar = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/display_url")
            .clickable(true));
    public static UiObject permAllowBtn = mDevice.findObject(new UiSelector()
            .resourceId("com.android.packageinstaller:id/permission_allow_button")
            .clickable(true));
    public static UiObject inlineAutocompleteEditText = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/urlView")
            .focused(true)
            .enabled(true));
    public static UiObject searchSuggestionsTitle = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/enable_search_suggestions_title")
            .enabled(true));
    public static UiObject searchSuggestionsButtonYes = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/enable_search_suggestions_button")
            .enabled(true));
    public static UiObject cleartextField = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/clearView")
            .enabled(true));
    public static UiObject hint = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/searchView")
            .clickable(true));
    public static UiObject suggestionList = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/suggestionList"));
    public static UiObject suggestion = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/suggestion"));
    public static UiObject webView = mDevice.findObject(new UiSelector()
            .className("android.webkit.WebView")
            .enabled(true));
    public static UiObject geckoView = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/webview")
            .enabled(true));
    public static UiObject progressBar = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/progress")
            .enabled(true));
    public static UiObject tryAgainBtn = mDevice.findObject(new UiSelector()
            .resourceId("errorTryAgain")
            .clickable(true));
    public static ViewInteraction floatingEraseButton = onView(
            allOf(withId(R.id.erase), isDisplayed()));
    public static UiObject browserViewSettingsMenuItem = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/settings")
            .clickable(true));
    public static UiObject erasedMsg = TestHelper.mDevice.findObject(new UiSelector()
            .text("Your browsing history has been erased.")
            .resourceId(getAppName() + ":id/snackbar_text")
            .enabled(true));
    public static UiObject lockIcon = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/lock")
            .description("Secure connection"));
    public static UiObject notificationBarDeleteItem = TestHelper.mDevice.findObject(new UiSelector()
            .text("Erase browsing history")
            .resourceId("android:id/text")
            .enabled(true));
    public static UiObject notificationExpandSwitch = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId("android:id/expand_button")
            .packageName(getAppName() + "")
            .enabled(true));
    public static UiObject notificationOpenItem = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId("android:id/action0")
            .description("Open")
            .enabled(true));
    public static UiObject notificationEraseOpenItem = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId("android:id/action0")
            .description("Erase and Open")
            .enabled(true));
    public static UiObject blockOffIcon = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/block")
            .enabled(true));
    public static UiObject AddtoHSmenuItem = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/add_to_homescreen")
            .enabled(true));
    public static UiObject AddtoHSCancelBtn = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/addtohomescreen_dialog_cancel")
            .enabled(true));
    public static UiObject AddtoHSOKBtn = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/addtohomescreen_dialog_add")
            .enabled(true));
    public static UiObject AddautoBtn = TestHelper.mDevice.findObject(new UiSelector()
            .className("android.widget.Button")
            .instance(1)
            .enabled(true));
    public static UiObject shortcutTitle = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/edit_title")
            .enabled(true));
    public static UiObject savedNotification = TestHelper.mDevice.findObject(new UiSelector()
            .text("Download complete.")
            .resourceId("android:id/text")
            .enabled(true));

    public static UiObject securityInfoIcon = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId(TestHelper.getAppName() + ":id/security_info")
            .enabled(true));
    public static UiObject identityState = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId(TestHelper.getAppName() + ":id/site_identity_state")
            .enabled(true));

    public static UiObject downloadTitle = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId(TestHelper.getAppName() + ":id/title_template")
            .enabled(true));
    public static UiObject downloadFileName = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId(TestHelper.getAppName() + ":id/download_dialog_file_name")
            .enabled(true));
    public static UiObject downloadWarning = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId(TestHelper.getAppName() + ":id/download_dialog_warning")
            .enabled(true));
    public static UiObject downloadCancelBtn = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId(TestHelper.getAppName() + ":id/download_dialog_cancel")
            .enabled(true));
    public static UiObject downloadBtn = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId(TestHelper.getAppName() + ":id/download_dialog_download")
            .enabled(true));
    public static UiObject completedMsg = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId(TestHelper.getAppName() + ":id/snackbar_text")
            .enabled(true));

    /********* Main View Menu Item Locators ***********/
    public static UiObject whatsNewItem = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/whats_new")
            .enabled(true));
    public static UiObject HelpItem = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/help")
            .enabled(true));
    public static UiObject settingsMenuItem = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/settings")
            .enabled(true));
    public static UiObject blockCounterItem = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/trackers_count"));
    public static UiObject menulist = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/list")
            .enabled(true));
    public static String getMenuItemText(UiObject item) throws UiObjectNotFoundException {
        return item.getChild(new UiSelector().index(0))
                .getChild(new UiSelector().index(0)).getText();
    }

    /********** Share Menu Dialog ********************/
    public static UiObject shareMenuHeader = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId("android:id/title")
            .text("Share via")
            .enabled(true));
    public static UiObject shareAppList = TestHelper.mDevice.findObject(new UiSelector()
            .resourceId("android:id/resolver_list")
            .enabled(true));

    /********* Settings Menu Item Locators ***********/
    public static UiObject settingsMenu = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/recycler_view"));
    public static UiObject settingsHeading = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/toolbar")
            .enabled(true));
    public static UiObject navigateUp = mDevice.findObject(new UiSelector()
            .description("Navigate up"));
    public static UiObject toggleAnalyticBlock = mDevice.findObject(new UiSelector()
            .className("android.widget.Switch")
            .instance(1));
    public static UiObject refreshBtn = mDevice.findObject(new UiSelector()
            .resourceId(getAppName() + ":id/refresh")
            .enabled(true));

    public static void expandNotification() throws UiObjectNotFoundException {
        if (!notificationOpenItem.waitForExists(waitingTime)) {
            if (!notificationExpandSwitch.exists()) {
                notificationBarDeleteItem.pinchOut(50, 5);
            } else {
                notificationExpandSwitch.click();
            }
            assertTrue(notificationOpenItem.exists());
        }
    }

    public static final int X_OFFSET = 20;
    public static final int Y_OFFSET = 500;
    public static final int STEPS = 10;

    private static DisplayMetrics devicePixels() {
        DisplayMetrics metrics = Resources.getSystem().getDisplayMetrics();
        return metrics;
    }

    public static void swipeScreenLeft() {
        DisplayMetrics metrics = devicePixels();
        mDevice.swipe(metrics.widthPixels - X_OFFSET, Y_OFFSET, 0, Y_OFFSET, STEPS);
    }

    public static void swipeScreenRight() {
        DisplayMetrics metrics = devicePixels();
        mDevice.swipe(X_OFFSET, Y_OFFSET, metrics.widthPixels, Y_OFFSET, STEPS);
    }

    public static void waitForIdle() {
        mDevice.waitForIdle(waitingTime);
    }
    public static void pressEnterKey() {
        mDevice.pressKeyCode(KEYCODE_ENTER);
    }
    public static void pressBackKey() {
        mDevice.pressBack();
    }
    public static void pressHomeKey() {
        mDevice.pressHome();
    }
    public static void openNotification() {
        mDevice.openNotification();
    }

    public static MockResponse createMockResponseFromAsset(@NonNull String fileName) throws IOException {
        return new MockResponse()
                .setBody(TestHelper.readTestAsset(fileName));
    }

    public static Buffer readTestAsset(String filename) throws IOException {
        try (final InputStream stream = InstrumentationRegistry.getContext().getAssets().open(filename)) {
            return readStreamFile(stream);
        }
    }

    public static Buffer readStreamFile(InputStream file) throws IOException {

        Buffer buffer = new Buffer();
        buffer.writeAll(Okio.source(file));
        return buffer;
    }

    public static  String readFileToString(File file) throws IOException {
        System.out.println("Reading file: " + file.getAbsolutePath());

        try (final FileInputStream stream = new FileInputStream(file)) {
            return readStreamIntoString(stream);
        }
    }

    public static  String readStreamIntoString(InputStream stream) throws IOException {
        try (final BufferedReader reader = new BufferedReader(
                new InputStreamReader(stream, StandardCharsets.UTF_8))) {

            final StringBuilder builder = new StringBuilder();
            String line;

            while ((line = reader.readLine()) != null) {
                builder.append(line);
            }
            reader.close();

            return builder.toString();
        }
    }

    public static void waitForWebSiteTitleLoad() {
        onWebView(withText("focus test page"));
    }

    public static void selectGeckoForKlar() {
        InstrumentationRegistry.getTargetContext().getSharedPreferences("mozilla.components.service.fretboard.overrides",
                Context.MODE_PRIVATE)
                .edit()
                .putBoolean("use-gecko", AppConstants.INSTANCE.isKlarBuild())
                .commit();
    }
}

File: mobile/android/focus-android/app/src/androidTest/java/org/mozilla/focus/helpers/TestHelper.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
package org.mozilla.focus.helpers

import android.app.PendingIntent
import android.app.UiAutomation
import android.content.ActivityNotFoundException
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.os.Build
import android.os.Build.VERSION.SDK_INT
import android.os.storage.StorageManager
import android.os.storage.StorageVolume
import android.util.Log
import android.view.KeyEvent
import android.view.inputmethod.InputMethodManager
import androidx.browser.customtabs.CustomTabColorSchemeParams
import androidx.browser.customtabs.CustomTabsIntent
import androidx.browser.customtabs.CustomTabsIntent.SHARE_STATE_ON
import androidx.core.net.toUri
import androidx.test.espresso.Espresso
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.intent.Intents.intended
import androidx.test.espresso.intent.matcher.IntentMatchers.toPackage
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.espresso.matcher.ViewMatchers.withText
import androidx.test.espresso.web.sugar.Web
import androidx.test.filters.SdkSuppress
import androidx.test.platform.app.InstrumentationRegistry.getInstrumentation
import androidx.test.uiautomator.UiDevice
import androidx.test.uiautomator.UiSelector
import junit.framework.AssertionFailedError
import mozilla.components.support.utils.ext.getApplicationInfoCompat
import okio.Buffer
import org.hamcrest.Matchers
import org.hamcrest.Matchers.allOf
import org.junit.Assert
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.mozilla.focus.R
import org.mozilla.focus.activity.IntentReceiverActivity
import org.mozilla.focus.utils.IntentUtils
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.util.concurrent.TimeUnit

@Suppress("TooManyFunctions")
object TestHelper {
    @JvmField
    var mDevice = UiDevice.getInstance(getInstrumentation())
    val waitingTime = TimeUnit.SECONDS.toMillis(15)
    val pageLoadingTime = TimeUnit.SECONDS.toMillis(25)
    val waitingTimeShort: Long = TimeUnit.SECONDS.toMillis(3)

    private val charPool: List<Char> = ('a'..'z') + ('A'..'Z') + ('0'..'9')
    fun randomString(stringLength: Int) =
        (1..stringLength)
            .map { kotlin.random.Random.nextInt(0, charPool.size) }
            .map(charPool::get)
            .joinToString("")

    @JvmStatic
    val getTargetContext: Context = getInstrumentation().targetContext

    @JvmStatic
    val packageName: String = getTargetContext.packageName

    @JvmStatic
    val appName: String = getTargetContext.getString(R.string.app_name)

    fun getStringResource(id: Int) = getTargetContext.resources.getString(id, appName)

    fun verifySnackBarText(text: String) {
        val snackbarText = mDevice.findObject(UiSelector().textContains(text))
        assertTrue(snackbarText.waitForExists(waitingTime))
    }

    fun clickSnackBarActionButton(action: String) {
        val snackbarActionButton =
            onView(
                allOf(
                    withId(R.id.snackbar_action),
                    withText(action),
                ),
            )
        snackbarActionButton.perform(click())
    }

    fun waitUntilSnackBarGone() {
        mDevice.findObject(UiSelector().resourceId("$appName:id/snackbar_layout"))
            .waitUntilGone(waitingTime)
    }

    fun isPackageInstalled(packageName: String): Boolean {
        return try {
            val packageManager = getInstrumentation().context.packageManager
            packageManager.getApplicationInfoCompat(packageName, 0).enabled
        } catch (exception: PackageManager.NameNotFoundException) {
            Log.d("TestLog", exception.message.toString())
            false
        }
    }

    fun restartApp(activity: MainActivityFirstrunTestRule) {
        with(activity) {
            finishActivity()
            mDevice.waitForIdle()
            launchActivity(null)
        }
    }

    // exit to the main view
    fun exitToTop() {
        val homeScreen =
            mDevice.findObject(UiSelector().resourceId("$packageName:id/landingLayout"))
        var homeScreenVisible = false
        while (!homeScreenVisible) {
            mDevice.pressBack()
            homeScreenVisible = homeScreen.waitForExists(2000)
        }
    }

    // exit to the browser view
    fun exitToBrowser() {
        val browserScreen =
            mDevice.findObject(UiSelector().resourceId("$packageName:id/main_content"))
        var browserScreenVisible = false
        while (!browserScreenVisible) {
            mDevice.pressBack()
            browserScreenVisible = browserScreen.waitForExists(2000)
        }
    }

    fun setNetworkEnabled(enabled: Boolean) {
        when (enabled) {
            true -> {
                mDevice.executeShellCommand("svc data enable")
                mDevice.executeShellCommand("svc wifi enable")
            }

            false -> {
                mDevice.executeShellCommand("svc data disable")
                mDevice.executeShellCommand("svc wifi disable")
            }
        }
        mDevice.waitForIdle(waitingTime)
    }

    // verifies localized strings in different UIs
    fun verifyTranslatedTextExists(text: String) =
        assertTrue(mDevice.findObject(UiSelector().text(text)).waitForExists(waitingTime))

    fun openAppFromExternalLink(url: String) {
        val intent = Intent().apply {
            action = Intent.ACTION_VIEW
            data = url.toUri()
            `package` = packageName
            flags = Intent.FLAG_ACTIVITY_NEW_TASK
        }
        try {
            getTargetContext.startActivity(intent)
        } catch (ex: ActivityNotFoundException) {
            intent.setPackage(null)
            getTargetContext.startActivity(intent)
        }
    }

    @SdkSuppress(minSdkVersion = Build.VERSION_CODES.R)
    fun verifyDownloadedFileOnStorage(fileName: String) {
        val storageManager =
            getInstrumentation().targetContext.getSystemService(Context.STORAGE_SERVICE) as StorageManager
        val storageVolumes = storageManager.storageVolumes
        val storageVolume: StorageVolume = storageVolumes[0]
        val file = File("${storageVolume.directory!!.path}/Download/$fileName")
        assertTrue(file.exists())
    }

    // Method for granting app permission to access location/camera/mic
    fun grantAppPermission() {
        if (SDK_INT >= 23) {
            mDevice.findObject(
                UiSelector().textContains(
                    when (SDK_INT) {
                        Build.VERSION_CODES.R ->
                            "While using the app"
                        else -> "Allow"
                    },
                ),
            ).click()
        }
    }

    fun UiAutomation.executeShellCommandBlocking(command: String) {
        val output = executeShellCommand(command)
        FileInputStream(output.fileDescriptor).use { it.readBytes() }
    }

    @JvmStatic
    fun pressEnterKey() {
        mDevice.pressKeyCode(KeyEvent.KEYCODE_ENTER)
    }

    @JvmStatic
    fun pressBackKey() {
        mDevice.pressBack()
    }

    @JvmStatic
    fun pressHomeKey() {
        mDevice.pressHome()
    }

    fun createCustomTabIntent(
        pageUrl: String,
        customMenuItemLabel: String = "",
        customActionButtonDescription: String = "",
    ): Intent {
        val appContext = getInstrumentation()
            .targetContext
            .applicationContext
        val pendingIntent = PendingIntent.getActivity(appContext, 0, Intent(), IntentUtils.defaultIntentPendingFlags())

        val customTabColorSchemeBuilder = CustomTabColorSchemeParams.Builder()
        customTabColorSchemeBuilder.setToolbarColor(Color.MAGENTA)

        val customTabsIntent = CustomTabsIntent.Builder()
            .addMenuItem(customMenuItemLabel, pendingIntent)
            .setShareState(SHARE_STATE_ON)
            .setActionButton(createTestBitmap(), customActionButtonDescription, pendingIntent, true)
            .setDefaultColorSchemeParams(customTabColorSchemeBuilder.build())
            .build()
        customTabsIntent.intent.data = pageUrl.toUri()
        customTabsIntent.intent.component = ComponentName(appContext, IntentReceiverActivity::class.java)
        return customTabsIntent.intent
    }

    fun assertNativeAppOpens(appPackageName: String) {
        try {
            if (isPackageInstalled(packageName)) {
                intended(toPackage(appPackageName))
            }
        } catch (e: AssertionFailedError) {
            e.printStackTrace()
        }
    }

    private fun createTestBitmap(): Bitmap {
        val bitmap = Bitmap.createBitmap(100, 100, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        canvas.drawColor(Color.GREEN)
        return bitmap
    }

    /**
     * Wrapper for tests to run only when certain conditions are met.
     * For example: this method will avoid accidentally running a test on GV versions where the feature is disabled.
     */
    fun runWithCondition(condition: Boolean, testBlock: () -> Unit) {
        if (condition) {
            testBlock()
        }
    }

    /********* Old code locators - used only in Screenshots tests  */
    // wait for web area to be visible
    @JvmStatic
    fun waitForWebContent() {
        Assert.assertTrue(geckoView.waitForExists(waitingTime))
    }

    @JvmField
    var menuButton = Espresso.onView(
        Matchers.allOf(
            ViewMatchers.withId(R.id.menuView),
            ViewMatchers.isDisplayed(),
        ),
    )

    @JvmField
    var permAllowBtn = mDevice.findObject(
        UiSelector()
            .textContains("Allow")
            .clickable(true),
    )

    @JvmField
    var webView = mDevice.findObject(
        UiSelector()
            .className("android.webkit.WebView")
            .enabled(true),
    )
    var geckoView = mDevice.findObject(
        UiSelector()
            .resourceId(packageName + ":id/engineView")
            .enabled(true),
    )

    @JvmField
    var progressBar = mDevice.findObject(
        UiSelector()
            .resourceId(packageName + ":id/progress")
            .enabled(true),
    )

    @JvmField
    var AddtoHSmenuItem = mDevice.findObject(
        UiSelector()
            .resourceId(packageName + ":id/add_to_homescreen")
            .enabled(true),
    )

    @JvmField
    var AddtoHSCancelBtn = mDevice.findObject(
        UiSelector()
            .resourceId(packageName + ":id/addtohomescreen_dialog_cancel")
            .enabled(true),
    )

    @JvmField
    var securityInfoIcon = mDevice.findObject(
        UiSelector()
            .resourceId(packageName + ":id/security_info")
            .enabled(true),
    )

    @JvmField
    var identityState = mDevice.findObject(
        UiSelector()
            .resourceId(packageName + ":id/site_identity_state")
            .enabled(true),
    )

    @JvmField
    var shareAppList = mDevice.findObject(
        UiSelector()
            .resourceId("android:id/resolver_list")
            .enabled(true),
    )

    @JvmStatic
    @Throws(IOException::class)
    fun readTestAsset(filename: String?): Buffer {
        getInstrumentation().getContext().assets.open(filename!!)
            .use { stream -> return readStreamFile(stream) }
    }

    @Throws(IOException::class)
    fun readStreamFile(file: InputStream?): Buffer {
        val buffer = Buffer()
        buffer.write(file!!.readBytes())
        return buffer
    }

    @JvmStatic
    fun waitForWebSiteTitleLoad() {
        Web.onWebView(ViewMatchers.withText("focus test page"))
    }

    @JvmStatic
    fun verifyKeyboardVisibility(isExpectedToBeVisible: Boolean) {
        val imm = getTargetContext.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager

        if (isExpectedToBeVisible) {
            assertTrue(imm.isAcceptingText)
        } else {
            assertFalse(imm.isAcceptingText)
        }
    }
}
